src/lsmr.jl

```diff
@@ -0,0 +1,250 @@
+## LSMR
+##
+## Minimize ||Ax-b||^2 + λ^2 ||x||^2
+##
+## Adapted from the BSD-licensed Matlab implementation at
+## http://web.stanford.edu/group/SOL/software/lsmr/
+##
+## A is a StridedVecOrMat or anything that implements 
+## A_mul_B!(α, A, b, β, c) updates c -> α Ab + βc
+## Ac_mul_B!(α, A, b, β, c) updates c -> α A'b + βc
+## eltype(A)
+## size(A)
+## (this includes SparseMatrixCSC)
+## x, v, h, hbar are AbstractVectors or anything that implements
+## norm(x)
+## copy!(x1, x2)
+## scale!(x, α)
+## axpy!(α, x1, x2)
+## similar(x, T)
+## length(x)
+## b is an AbstractVector or anything that implements
+## eltype(b)
+## norm(b)
+## copy!(x1, x2)
+## fill!(b, α)
+## scale!(b, α)
+## similar(b, T)
+## length(b)
+
+##############################################################################
+
+
+## Arguments:
+## x is initial x0. Transformed in place to the solution.
+## b equals initial b. Transformed in place
+## v, h, hbar are storage arrays of length size(A, 2)
+function lsmr!(x, A, b, v, h, hbar; 
+	atol::Number = 1e-6, btol::Number = 1e-6, conlim::Number = 1e8, 
+	maxiter::Integer = max(size(A,1), size(A,2)), λ::Number = 0)
+
+	# Sanity-checking
+	m = size(A, 1)
+	n = size(A, 2)
+	length(x) == n || error("x has length $(length(x)) but should have length $n")
+	length(v) == n || error("v has length $(length(v)) but should have length $n")
+	length(h) == n || error("h has length $(length(h)) but should have length $n")
+	length(hbar) == n || error("hbar has length $(length(hbar)) but should have length $n")
+	length(b) == m || error("b has length $(length(b)) but should have length $m")
+
+
+	T = Adivtype(A, b)
+	Tr = real(T)
+	normrs = Tr[]
+	normArs = Tr[]
+	conlim > 0 ? ctol = convert(Tr, inv(conlim)) : ctol = zero(Tr)
+	# form the first vectors u and v (satisfy  β*u = b,  α*v = A'u)
+	u = A_mul_B!(-1, A, x, 1, b)
+	β = norm(u)
+	β > 0 && scale!(u, inv(β))
+	Ac_mul_B!(1, A, u, 0, v)
+	α = norm(v)
+	α > 0 && scale!(v, inv(α))
+
+	# Initialize variables for 1st iteration.
+	ζbar = α * β
+	αbar = α
+	ρ = one(Tr)
+	ρbar = one(Tr)
+	cbar = one(Tr)
+	sbar = zero(Tr)
+
+	copy!(h, v)
+	fill!(hbar, zero(Tr))
+
+	# Initialize variables for estimation of ||r||.
+	βdd = β
+	βd = zero(Tr)
+	ρdold = one(Tr)
+	τtildeold = zero(Tr)
+	θtilde  = zero(Tr)
+	ζ = zero(Tr)
+	d = zero(Tr)
+
+	# Initialize variables for estimation of ||A|| and cond(A).
+	normA, condA, normx = -one(Tr), -one(Tr), -one(Tr)
+	normA2 = abs2(α)
+	maxrbar = zero(Tr)
+	minrbar = 1e100
+
+	# Items for use in stopping rules.
+	normb = β
+	istop = 0 
+	normr = β
+	normAr = α * β
+	tests = Tuple{Tr, Tr, Tr}[]
+	iter = 0
+	# Exit if b = 0 or A'b = 0.
+	if normAr != 0 
+		while iter < maxiter
+			iter += 1
+			A_mul_B!(1, A, v, -α, u)
+			β = norm(u)
+			if β > 0
+				scale!(u, inv(β))
+				Ac_mul_B!(1, A, u, -β, v)
+				α = norm(v)
+				α > 0 && scale!(v, inv(α))
+			end
+		
+			# Construct rotation Qhat_{k,2k+1}.
+			αhat = sqrt(abs2(αbar) + abs2(λ))
```

Comment:
Can we be sure that this quantity won't overflow, i.e. that `hypot` is not necessary?